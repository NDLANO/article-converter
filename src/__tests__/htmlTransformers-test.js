/**
 * Copyright (c) 2018-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import cheerio from 'cheerio';
import { prettify } from './testHelpers';
import {
  htmlTransforms,
  moveReactPortals,
  transformAsides,
} from '../htmlTransformers';

test('htmlTransforms changes ol to accommodate frontend styling', () => {
  const content = cheerio.load(`
  <section>
    <ol data-type='letters'>
        <li>Lorem ipsum dolor sit amet...</li>
    </ol>
  </section>
    <ol>
        <li>Lorem ipsum dolor sit amet...</li>
    </ol>
  </section>`);

  htmlTransforms.forEach(tagReplacer => tagReplacer(content));
  const result = content.html();

  expect(result).toMatch('<ol class="ol-list--roman">');
  expect(result).toMatch('<ol>');
});

test('htmlTransforms adds display block to math tags', () => {
  const content = cheerio.load(`
  <section>
    <math xmlns="http://www.w3.org/1998/Math/MathML">
      <mi>f</mi><mo>'</mo>
      <mfenced><mrow><mn>0</mn><mo>,</mo><mn>5</mn></mrow></mfenced>
      <mo>=</mo><mn>2</mn><mo>&#xB7;</mo><mn>0</mn><mo>,</mo><mn>5</mn><mo>=</mo><mn>1</mn>
    </math>
    <math xmlns="http://www.w3.org/1998/Math/MathML">
      <mn>0</mn><mo>,</mo><mn>5</mn>
    </math>
  </section>
  `);

  htmlTransforms.forEach(tagReplacer => tagReplacer(content));
  const result = content.html();

  expect(result).toMatchSnapshot();
});

test('htmlTransforms changes p to accommodate frontend styling', () => {
  const content = cheerio.load(`
  <section>
  <p data-align='center'>Lorem ipsum dolor sit amet...</p>
  </section>`);

  htmlTransforms.forEach(tagReplacer => tagReplacer(content));
  const result = content.html();

  expect(result).toMatch('<p class="u-text-center">');
});

test('htmlTransforms changes span with data-size to a span with font size', () => {
  const content = cheerio.load(`
  <section>
  Lorem ipsum <span data-size="large">dolor</span> sit amet...
  </section>`);

  htmlTransforms.forEach(tagReplacer => tagReplacer(content));
  const result = content.html();

  expect(result).toMatch(
    'Lorem ipsum <span class="u-large-body-text">dolor</span> sit amet'
  );
});

test('move react portals to bottom', () => {
  const content = cheerio.load(`
  <section>
    <figure>
      <p>Lorem ipsum dolor sit amet...</p>
      <div data-react-universal-portal="true"><h2>Modal dialog</h2><div>Stuff</div></div>
    </figure>
  </section>`);

  moveReactPortals(content);

  const result = content.html();

  expect(result).toMatchSnapshot();
});

test('transformAsides duplicates right column aside for better narrowscreen experience', () => {
  const content = cheerio.load(`
  <section>
    <figure>
      <p>Lorem ipsum dolor sit amet...</p>
    </figure>
    <aside><h2>Test1</h2><div>Stuff</div></aside>
    <p>sdfjljklsdfjlsdf</p>
    <figure>
      <p>Lorem ipsum dolor sit amet...</p>
      <div data-react-universal-portal="true"><h2>Modal dialog</h2><div>Stuff</div></div>
    </figure>
    <p>sdfjljklsdfjlsdf</p>
  </section>`);

  transformAsides(content);

  const result = content('body').html();
  expect(prettify(result)).toMatchSnapshot();
});

test('transformAsides encloses aside duplication to sections', () => {
  const content = cheerio.load(`
  <section>
    <aside><h2>Test1</h2><div>Stuff</div></aside>
    <p>Lorem</p>
  </section>
  <section>
    <aside><h2>Test1</h2><div>Stuff</div></aside>
    <p>Ipsum</p>
  </section>`);

  transformAsides(content);

  const result = content('body').html();
  expect(prettify(result)).toMatchSnapshot();
});

test('transformAsides replaces factAsides with factbox component', () => {
  const content = cheerio.load(`
  <section>
    <p>Lorem ipsum dolor sit amet...</p>
    <aside data-type='factAside'><h2>Test1</h2><div>Stuff</div></aside>
  </section>
    <aside data-type="factAside"><h2>Test2</h2><div>Other stuff</div></aside>
    <p>Lorem ipsum dolor sit amet...</p>
  </section>`);

  transformAsides(content);

  const result = content('body').html();
  expect(prettify(result)).toMatchSnapshot();
});
