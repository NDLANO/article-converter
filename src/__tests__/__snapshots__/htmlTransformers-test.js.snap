// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`htmlTransforms adds display block to math tags 1`] = `
"<html><head></head><body><section>
    <math xmlns=\\"http://www.w3.org/1998/Math/MathML\\" display=\\"block\\">
      <mi>f</mi><mo>&apos;</mo>
      <mfenced><mrow><mn>0</mn><mo>,</mo><mn>5</mn></mrow></mfenced>
      <mo>=</mo><mn>2</mn><mo>&#xB7;</mo><mn>0</mn><mo>,</mo><mn>5</mn><mo>=</mo><mn>1</mn>
    </math>
    <math xmlns=\\"http://www.w3.org/1998/Math/MathML\\" display=\\"block\\">
      <mn>0</mn><mo>,</mo><mn>5</mn>
    </math>
  </section>
  </body></html>"
`;

exports[`move react portals to bottom 1`] = `
"<html><head></head><body><section>
    <figure>
      <p>Lorem ipsum dolor sit amet...</p>
      
    </figure>
  </section><div data-react-universal-portal=\\"true\\"><h2>Modal dialog</h2><div>Stuff</div></div></body></html>"
`;

exports[`transformAsides duplicates right column aside for better narrowscreen experience 1`] = `
"<section>
  <figure>
    <p>Lorem ipsum dolor sit amet...</p>
  </figure>
  <aside class=\\"c-aside c-aside--wideScreen\\">
    <div class=\\"c-aside__content\\">
      <h2>Test1</h2>
      <div>Stuff</div>
    </div>
  </aside>
  <p>sdfjljklsdfjlsdf</p>
  <figure>
    <p>Lorem ipsum dolor sit amet...</p>
    <div data-react-universal-portal=\\"true\\">
      <h2>Modal dialog</h2>
      <div>Stuff</div>
    </div>
  </figure>
  <p>sdfjljklsdfjlsdf</p>
  <aside class=\\"c-aside c-aside--narrowScreen\\">
    <div class=\\"c-aside__content\\">
      <h2>Test1</h2>
      <div>Stuff</div>
    </div>
  </aside>
</section>
"
`;

exports[`transformAsides encloses aside duplication to sections 1`] = `
"<section>
  <aside class=\\"c-aside c-aside--wideScreen\\">
    <div class=\\"c-aside__content\\">
      <h2>Test1</h2>
      <div>Stuff</div>
    </div>
  </aside>
  <p>Lorem</p>
  <aside class=\\"c-aside c-aside--narrowScreen\\">
    <div class=\\"c-aside__content\\">
      <h2>Test1</h2>
      <div>Stuff</div>
    </div>
  </aside>
</section>
<section>
  <aside class=\\"c-aside c-aside--wideScreen\\">
    <div class=\\"c-aside__content\\">
      <h2>Test1</h2>
      <div>Stuff</div>
    </div>
  </aside>
  <p>Ipsum</p>
  <aside class=\\"c-aside c-aside--narrowScreen\\">
    <div class=\\"c-aside__content\\">
      <h2>Test1</h2>
      <div>Stuff</div>
    </div>
  </aside>
</section>
"
`;

exports[`transformAsides replaces factAsides with factbox component 1`] = `
"<section>
  <p>Lorem ipsum dolor sit amet...</p>
  <aside class=\\"c-factbox\\">
    <div class=\\"c-factbox__content\\">
      <h2>Test1</h2>
      <div>Stuff</div>
    </div>
    <style data-emotion-css=\\"12v9xyl-StyledButton-buttonStyle-ButtonStyles\\">
      .css-12v9xyl-StyledButton-buttonStyle-ButtonStyles {
        display: inline-block;
        color: #ffffff;
        background-color: #20588f;
        border: 2px solid #20588f;
        border-radius: 4px;
        padding: 4px 13px;
        outline-width: 0;
        cursor: pointer;
        -webkit-text-decoration: none;
        text-decoration: none;
        font-size: 16px;
        font-size: 0.8888888888888888rem;
        line-height: 1.625;
        font-family: \\"Source Sans Pro\\", Helvetica, Arial, STKaiti, \\"华文楷体\\",
          KaiTi, SimKai, \\"楷体\\", KaiU, DFKai-SB, \\"標楷體\\", SongTi, \\"宋体\\",
          sans-serif;
        font-weight: 700;
        -webkit-transition: all 0.2s cubic-bezier(0.17, 0.04, 0.03, 0.94);
        transition: all 0.2s cubic-bezier(0.17, 0.04, 0.03, 0.94);
        box-shadow: none;
        text-align: center;
      }
      .css-12v9xyl-StyledButton-buttonStyle-ButtonStyles:hover,
      .css-12v9xyl-StyledButton-buttonStyle-ButtonStyles:focus {
        color: white;
        background-color: #184673;
        border: 2px solid #184673;
      }
      .css-12v9xyl-StyledButton-buttonStyle-ButtonStyles[disabled] {
        color: #777777;
        background-color: #e6e6e6;
        border-color: transparent;
        cursor: not-allowed;
      }
      .css-12v9xyl-StyledButton-buttonStyle-ButtonStyles:focus {
        box-shadow: 0 0 2px #20588f;
      }</style
    ><button
      type=\\"button\\"
      class=\\"c-factbox__button css-12v9xyl-StyledButton-buttonStyle-ButtonStyles e1ljv5qk0\\"
    ></button>
  </aside>
</section>
<aside class=\\"c-factbox\\">
  <div class=\\"c-factbox__content\\">
    <h2>Test2</h2>
    <div>Other stuff</div>
  </div>
  <style data-emotion-css=\\"12v9xyl-StyledButton-buttonStyle-ButtonStyles\\">
    .css-12v9xyl-StyledButton-buttonStyle-ButtonStyles {
      display: inline-block;
      color: #ffffff;
      background-color: #20588f;
      border: 2px solid #20588f;
      border-radius: 4px;
      padding: 4px 13px;
      outline-width: 0;
      cursor: pointer;
      -webkit-text-decoration: none;
      text-decoration: none;
      font-size: 16px;
      font-size: 0.8888888888888888rem;
      line-height: 1.625;
      font-family: \\"Source Sans Pro\\", Helvetica, Arial, STKaiti, \\"华文楷体\\",
        KaiTi, SimKai, \\"楷体\\", KaiU, DFKai-SB, \\"標楷體\\", SongTi, \\"宋体\\",
        sans-serif;
      font-weight: 700;
      -webkit-transition: all 0.2s cubic-bezier(0.17, 0.04, 0.03, 0.94);
      transition: all 0.2s cubic-bezier(0.17, 0.04, 0.03, 0.94);
      box-shadow: none;
      text-align: center;
    }
    .css-12v9xyl-StyledButton-buttonStyle-ButtonStyles:hover,
    .css-12v9xyl-StyledButton-buttonStyle-ButtonStyles:focus {
      color: white;
      background-color: #184673;
      border: 2px solid #184673;
    }
    .css-12v9xyl-StyledButton-buttonStyle-ButtonStyles[disabled] {
      color: #777777;
      background-color: #e6e6e6;
      border-color: transparent;
      cursor: not-allowed;
    }
    .css-12v9xyl-StyledButton-buttonStyle-ButtonStyles:focus {
      box-shadow: 0 0 2px #20588f;
    }</style
  ><button
    type=\\"button\\"
    class=\\"c-factbox__button css-12v9xyl-StyledButton-buttonStyle-ButtonStyles e1ljv5qk0\\"
  ></button>
</aside>
<p>Lorem ipsum dolor sit amet...</p>
"
`;

exports[`transformTables replaces table with <Table> 1`] = `
"<div class=\\"c-table__wrapper\\">
  <div class=\\"c-table__content\\">
    <div class=\\"c-table__left-shadow\\"></div>
    <table id=\\"testid\\" class=\\"c-table o-table\\">
      <thead>
        <tr>
          <th>test</th>
          <th>test</th>
          <th>test</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>test</td>
          <td>test</td>
          <td>test</td>
        </tr>
        <tr>
          <td>test</td>
          <td>test</td>
          <td>test</td>
        </tr>
        <tr>
          <td>test</td>
          <td>test</td>
          <td>test</td>
        </tr>
      </tbody>
    </table>
    <button
      type=\\"button\\"
      data-dialog-trigger-id=\\"dialog-testid\\"
      data-dialog-source-id=\\"testid\\"
      data-table-id=\\"testid\\"
      class=\\"c-table__expand-button\\"
      aria-label=\\"Vis stor versjon\\"
    >
      <svg
        fill=\\"currentColor\\"
        preserveAspectRatio=\\"xMidYMid meet\\"
        height=\\"1em\\"
        width=\\"1em\\"
        class=\\"c-icon\\"
        viewBox=\\"0 0 24 24\\"
        data-license=\\"Apache License 2.0\\"
        data-source=\\"Material Design\\"
        style=\\"vertical-align:middle\\"
      >
        <title>ZoomOutMap</title>
        <g>
          <path
            d=\\"M15 3l2.3 2.3-2.89 2.87 1.42 1.42L18.7 6.7 21 9V3zM3 9l2.3-2.3 2.87 2.89 1.42-1.42L6.7 5.3 9 3H3zm6 12l-2.3-2.3 2.89-2.87-1.42-1.42L5.3 17.3 3 15v6zm12-6l-2.3 2.3-2.87-2.89-1.42 1.42 2.89 2.87L15 21h6z\\"
            clip-path=\\"url(#b)\\"
          />
          <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\" clip-path=\\"url(#b)\\" />
        </g>
      </svg>
    </button>
    <div class=\\"c-table__right-shadow\\"></div>
  </div>
  <div data-react-universal-portal=\\"true\\">
    <div
      class=\\"c-dialog c-dialog--fullscreen\\"
      data-dialog-id=\\"dialog-testid\\"
      role=\\"dialog\\"
      aria-hidden=\\"true\\"
      aria-label=\\"Vis stor versjon\\"
    >
      <div class=\\"c-dialog__content\\">
        <button class=\\"c-dialog__close\\" type=\\"button\\">Lukk</button>
        <div class=\\"c-table__dialog\\">
          <div class=\\"c-table__expanded-table-wrapper\\">
            <div class=\\"c-table__content\\"></div>
          </div>
        </div>
      </div>
      <div class=\\"o-backdrop\\"></div>
    </div>
  </div>
</div>
"
`;
